### ISSUE TITLE
spdk iscsi tgt multi sessions error

### ISSUE THREAD

**Issue by zhangsaisai** [spdk/spdk#1427]
1、Use SPDK 19.01 and DPDK 18.11. Run SPDK iscsi_tgt with the config:
[iSCSI]
  NodeBase "iqn.2016-06.io.spdk"
  AuthFile /usr/local/etc/spdk/auth.conf
  MinConnectionsPerCore 4
  MaxSessions 1000
[PortalGroup1]
  Portal DA2 100.162.210.3:3260
[InitiatorGroup1]
  InitiatorName ANY
  Netmask ANY
[AIO]
  AIO /mnt/dev1/file1 AIO0 512
[TargetNode1]
  TargetName disk1
  TargetAlias "Data Disk1"
  Mapping PortalGroup1 InitiatorGroup1
  AuthMethod Auto
  AuthGroup AuthGroup1
  UseDigest Auto
  LUN0 AIO0

2、Configure initiator and discover target，then we want to test multi sessions on one target, so 
run iscsiadm as:
iscsiadm -m node -o update -n node.session.nr_sessions -v 300
iscsiadm -m node --login

Then we can see error below:
reactor.c:_spdk_reactor_run: *NOTICE*: Reactor started on core 0
conn.c:spdk_iscsi_drop_conns: *ERROR*: drop old connections iqn.2016-06.io.spdk:disk1 by iqn.1994-05.com.redhat:fd43e34480ef,i,0x00023ded0000
conn.c:spdk_iscsi_drop_conns: *ERROR*: exiting conn by iqn.1994-05.com.redhat:fd43e34480ef,i,0x00023ded0000 (11.158.213.9)
conn.c:spdk_iscsi_drop_conns: *ERROR*: exiting 1 conns
conn.c:spdk_iscsi_drop_conns: *ERROR*: drop old connections iqn.2016-06.io.spdk:disk1 by iqn.1994-05.com.redhat:fd43e34480ef,i,0x00023ded0000
conn.c:spdk_iscsi_drop_conns: *ERROR*: exiting conn by iqn.1994-05.com.redhat:fd43e34480ef,i,0x00023ded0000 (11.158.213.9)
conn.c: spdk_iscsi_drop_conns: *ERROR*: exiting 1 conns
conn.c: spdk_iscsi_drop_conns: *ERROR*: drop old connections iqn.2016-06.io.spdk:disk1 by iqn.1994-05.com.redhat:fd43e34480ef,i,0x00023ded0000

3、If we set iscsiadm session 200, as below:
iscsiadm -m node -o update -n node.session.nr_sessions -v 200
the spdk iscsi tgt not report error.

4、We don’t understand why the error is reported when node.session.nr_sessions sets 300, 
Can you help us see this problem together？ Thanks

**Comment by wanghailiangx** [comment]
Could you try again with the latest SPDK 20.04 and DPDK 19.11.0?

**Comment by wanghailiangx** [comment]
Unfortunately, I can't reproduce this issue, I tried with 3 pairs of versions: 
SPDK v19.01 / DPDK 18.08.0
SPDK v19.04 / DPDK 19.02.0
SPDK v20.07 / DPDK 19.11.0

Log as below:
```
Logging in to [iface: default, target: iqn.2016-06.io.spdk:disk1, portal: 192.168.11.11,3260] (multiple)
Logging in to [iface: default, target: iqn.2016-06.io.spdk:disk1, portal: 192.168.11.11,3260] (multiple)
Logging in to [iface: default, target: iqn.2016-06.io.spdk:disk1, portal: 192.168.11.11,3260] (multiple)
Logging in to [iface: default, target: iqn.2016-06.io.spdk:disk1, portal: 192.168.11.11,3260] (multiple)
Logging in to [iface: default, target: iqn.2016-06.io.spdk:disk1, portal: 192.168.11.11,3260] (multiple)
Logging in to [iface: default, target: iqn.2016-06.io.spdk:disk1, portal: 192.168.11.11,3260] (multiple)
Login to [iface: default, target: iqn.2016-06.io.spdk:disk1, portal: 192.168.11.11,3260] successful.
Login to [iface: default, target: iqn.2016-06.io.spdk:disk1, portal: 192.168.11.11,3260] successful.
Login to [iface: default, target: iqn.2016-06.io.spdk:disk1, portal: 192.168.11.11,3260] successful.
Login to [iface: default, target: iqn.2016-06.io.spdk:disk1, portal: 192.168.11.11,3260] successful.
Login to [iface: default, target: iqn.2016-06.io.spdk:disk1, portal: 192.168.11.11,3260] successful.
Login to [iface: default, target: iqn.2016-06.io.spdk:disk1, portal: 192.168.11.11,3260] successful.
```

My steps:
1.  dd if=/dev/zero of="/tmp/sample_aio" bs=512 count=500  (bs=4096 count=5000 is also tried)
2.  ./app/iscsi_tgt/iscsi_tgt -c iscsi.conf.in
([AIO]
  AIO /tmp/sample_aio AIO0 512
)
3.  iscsiadm -m node -o update -n node.session.nr_sessions -v 300  (even -v 500 is ok)
4. iscsiadm -m node --login
(My OS: kernel 4.13.10, Fedora release 25 )

**Comment by shuhei-matsumoto** [comment]
Hi @wanghailiangx and all,

Our team reported to me that probably the same issue occurred by SPDK 20.01.
I will investigate this issue too.

-Shuhei

**Comment by shuhei-matsumoto** [comment]
I don't think this is related but share this for just in case.
https://github.com/open-iscsi/open-iscsi/issues/108
Does this issue depend on open iscsi version?

**Comment by shuhei-matsumoto** [comment]
My colleague said that in his case, the host detected multipath failure but session was not released and then the session pool became out of resource.

@zhangsaisai 

The following was included in your log?
: iscsi.c: 562:create_iscsi_sess: *ERROR*: MaxSessions set to 128
: iscsi.c:1780:iscsi_op_login_set_conn_info: *ERROR*: create_sess() failed

If no, your issue and the issue my colleague got may be different.

**Comment by jimharris** [comment]
Hi @zhangsaisai,

Can you try to reproduce with SPDK v20.04 and/or latest master?  That would help us know if the issue still exists or was fixed in one of the releases since SPDK v19.01.

Thanks,

Jim

**Comment by zhangsaisai** [comment]
> Hi @zhangsaisai,
> 
> Can you try to reproduce with SPDK v20.04 and/or latest master? That would help us know if the issue still exists or was fixed in one of the releases since SPDK v19.01.
> 
> Thanks,
> 
> Jim

Hi, @jimharris，
We have investigated this problem about "drop old connections" when node.session.nr_sessions is 300. 
According to the iscsi protocol, a session between an iSCSI Initiator and iSCSI Target is defined by a session ID (SSID). The session ID is an element composed of an initiator part (ISID) and a target part (Target Portal Group Tag) group. The ISID is explicitly specified by the initiator when the session is established.

We use an old version of  Open-iSCSI tool as initiator. In iscsid daemon process(Open-iSCSI tool), ISID is a six-element array": uint8_t isid[6]. The ISID first three elements are fixed, and the fourth element are obtained based on session->id, and the last two elements are always 0.

	session->isid[0] = DRIVER_ISID_0;//(0x00)
	session->isid[1] = DRIVER_ISID_1;//(0x02)
	session->isid[2] = DRIVER_ISID_2;//(0x3D)

	session->isid[3] = session->id;

For different sessions, their isid is only one byte different (session->isid[3]), which is obtained according to the session id. For a byte, its size can only be up to 256 (FF). **If node.session.nr_sessions exceeds 256, although their session id is different, there will be duplicate isid.**

For example:
session id: 759, isid: 0x00023df70000
![image](https://user-images.githubusercontent.com/7865361/86309387-3328bc00-bc4e-11ea-8d0f-5a815e55d6fd.png)
![image](https://user-images.githubusercontent.com/7865361/86309497-7c790b80-bc4e-11ea-8a15-003870f3188c.png)

session id: 503, isid: 0x00023df70000
![image](https://user-images.githubusercontent.com/7865361/86309548-97e41680-bc4e-11ea-9c09-8c9ac438be1d.png)

But for iSCSI Target, during login, the target side judges whether it is a duplicate session based on the initiator_name (based on isid) of the session and target information. If it is a duplicate session, it will be dropped. In the scenario where only one target is set, if the isid of the two sessions is the same, then the target thinks that the two sessions are duplicates and will go to drop session.

		xconn_match =
			drop_all ? **xconn->initiator_name** : spdk_scsi_port_get_name(xconn->initiator_port);

		if (!**strcasecmp(conn_match, xconn_match)** &&
		    conn->target == xconn->target) {

			if (num == 0) {
				/*
				 * Only print this message before we report the
				 *  first dropped connection.
				 */
				SPDK_ERRLOG("drop old connections %s by %s\n",
					    conn->target->name, conn_match);
			}


If node.session.nr_sessions exceeds 256, then there will be repeated isid, and then the target will drop session.

This problem ”ISID has only ever had 8-bits of uniqueness“ has been fixed in open-iscsi, the patch link is：
https://github.com/open-iscsi/open-iscsi/commit/4c1f2d90ef1c73e33d9f1e4ae9c206ffe015a8f9

After updating the open-iscsi version, node.session.nr_sessions exceeding 256 will not cause the drop session problem.

Thanks, all.
